### PR TITLE
canutils/slcan: Add SLCAN utility

### PR DESCRIPTION
**Pull request by PetervdPerk-NXP**
## Summary
Implementation of SLCAN (CAN over serial) UART protocol to convert SocketCAN interfaces to UART interfaces

## Impact
None, just a new application

## Testing
Verifed on UCANS32K146-01 converting both CAN buses to 2 seperate serial ports.

### CODE REVIEW

File: canutils/slcan/slcan.h [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+ */
+#define SLCAN_REC_FIFO_FULL    (1 << 0)
+#define SLCAN_SND_FIFO_FULL    (1 << 1)
+#define SLCAN_ERROR_WARN       (1 << 2)
+#define SLCAN_DATA_OVERRUN     (1 << 3)
+#define SLCAN_ERROR_PASSIVE    (1 << 5)
+#define SLCAN_ARBITRATION_LOST (1 << 6)
+#define SLCAN_BUS_ERROR        (1 << 7)
+
+/* T1401557F8f601000000▒0 */
```
Comment (davids5) [inline canutils/slcan/slcan.h:45]: Is this a test command?

**Review by davids5 — CHANGES_REQUESTED** [review]
@PetervdPerk-NXP - Nice.

File: canutils/slcan/slcan.c [hunk 2]
```diff
@@ -0,0 +1,504 @@
...
+                          /* return clear flags */
+
+                          write(fd, "F00\r", 4);
+                        }
+                      else if (buf[0] == 'O')
+                        {
+                          /* open CAN interface */
+
+                          mode = 1;
+                          printf("Open interface\n");
```
Comment (acassis) [inline canutils/slcan/slcan.c:313]: I think it should be debug_print() here!

File: canutils/slcan/slcan.c [hunk 3]
```diff
@@ -0,0 +1,504 @@
...
+                          strncpy(ifr.ifr_name, argv[1], IFNAMSIZ - 1);
+                          ifr.ifr_name[IFNAMSIZ - 1] = '\0';
+
+                          ifr.ifr_ifru.ifru_can_data.arbi_bitrate =
+                            canspeed / 1000; /* Convert bit/s to kbit/s */
+                          ifr.ifr_ifru.ifru_can_data.arbi_samplep = 80;
+
+                          if (ioctl(s, SIOCSCANBITRATE, &ifr) < 0)
+                            {
+                              printf("set speed %d failed\n", canspeed);
```
Comment (acassis) [inline canutils/slcan/slcan.c:366]: debug_print() too

File: canutils/slcan/slcan.c [hunk 4]
```diff
@@ -0,0 +1,504 @@
...
+                              fail_return(fd);
+                            }
+                          else
+                            {
+                              printf("set speed %d\n", canspeed);
```
Comment (acassis) [inline canutils/slcan/slcan.c:371]: here too: debug_print()

**Review by acassis — CHANGES_REQUESTED** [review]

File: canutils/slcan/slcan.c [hunk 5]
```diff
@@ -0,0 +1,504 @@
...
+#define DEFAULT_PRIORITY 100
+#define DEFAULT_STACK_SIZE 2048
+
+#ifdef CONFIG_SLCAN_TRACE
+#  define DEBUG 1
+#else
+#  define DEBUG 0
+#endif
+
+#define debug_print(fmt, ...) \
```
[the remark this answers is not in the document]
Comment (PetervdPerk-NXP) [reply canutils/slcan/slcan.c:67]: Sure no problem, modified now for use with syslog

File: canutils/slcan/slcan.h (hunk 1 above)
Comment (PetervdPerk-NXP) [reply canutils/slcan/slcan.h]: Indeed a leftover from testing, to verify this I would recommend using slcand on a Linux host http://manpages.ubuntu.com/manpages/cosmic/man1/slcand.1.html

File: canutils/slcan/slcan.c (hunk 2 above)
Comment (PetervdPerk-NXP) [reply canutils/slcan/slcan.c]: Thanks, I've cleaned up all the printouts now

**Review by davids5 — APPROVED** [review]
@PetervdPerk-NXP - thank you! 
LGTM
@acassis - Have all you comments been resolved? If so Please merge.

**Comment by PetervdPerk-NXP** [comment]
@acassis Anything still open to review?

**Comment by acassis** [comment]
I think it is fine. I will update my review

**Review by acassis — APPROVED** [review]

**Pull request merged**